PrestaShop/PrestaShop
# New BO page Experimental Feature

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Related to https://github.com/PrestaShop/PrestaShop/issues/23235, fixes https://github.com/PrestaShop/PrestaShop/issues/23998
| How to test?      | See below
| Possible impacts? | All the new files and settings introduced in this PR are related to the Experimental Features BO page. There should be no side-effects.

# Description

This Pull Request introduces a new Back Office page "Experimental Features" to manipulate feature flags.

It looks like this now
![Capture d’écran 2021-03-19 à 15 42 56](https://user-images.githubusercontent.com/3830050/111800775-946fe880-88cc-11eb-883d-fd056dedcefc.png)

## Behavior

For each feature flag registered in database, a switch is displayed. Feature flag label and help message are dynamic, fetched from database and translated.

When you hit "Submit" a confirmation window is displayed to confirm this is intentional.
![Capture d’écran 2021-03-19 à 15 43 10](https://user-images.githubusercontent.com/3830050/111800803-99cd3300-88cc-11eb-921e-c8bc72858a3e.png)

Some improvements in this confirmation window [are scheduled for another PR](https://github.com/PrestaShop/PrestaShop/issues/23235): it should not be displayed if merchant has only disabled features, it should use plural or singular depending if merchant has enabled one or many flags...

## Implementation

The form is built as a regular [Settings Form](https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/forms/settings-forms/) so it relies on
- a Controller
- a FormDataProvider in charge of fetching the data
- a Symfony form type
- twig form rendering

The above classes are not automatically tested because they are mainly boilerplate classes, the behavior is better tested on the application level.

Initially I wanted to use a CollectionType to handle the "there are as many switches as feature flags" logic but it seems CollectionType requires all children types to be identical. So I could not customize each Switch label or help text. This is why I used a basic foreach inside the Type, adding as many SwitchTypes as requested.

On the model side, everything happens in the FeatureFlagsModifier. It fetches the data, formats it, validates submitted payload and handles model state changes. It uses Doctrine. It is unit tested.

# How to test

First, you need to insert some feature flags in your database ! After installing a shop with the branch you must run the following SQL script
```
INSERT INTO `ps_feature_flag` (`id_feature_flag`, `name`, `state`, `description_wording`, `label_domain`, `label_wording`, `description_domain`)
VALUES
	(1, 'product_page_v2', 0, 'This is a helpful text', 's', 'Product Page V2', 's');

INSERT INTO `ps_feature_flag` (`id_feature_flag`, `name`, `state`, `description_wording`, `label_domain`, `label_wording`, `description_domain`)
VALUES
	(2, 'new_experimental_feature_page', 0, 'Le melon c est bon', 's', 'A fake Feature Flag For QA', 's');
```

Then you can open the new Back Office from the navigation navbar Configure > Advanced Parameters > Experimental Features. Now you can validate the above behavior. Don't hesitate to compare the state of the flags you see and the SQL data: the `state` field is a boolean, so when the flag is disabled it should have the SQL value 0 while when the flag is enabled it should have the value 1.

You can add more flags using the SQL script as an example.

### Incoming PRs following this one

Please note [that some more PRs will polish](https://github.com/PrestaShop/PrestaShop/issues/23235) this new Back Office page

#### When user installs 1.7.8.0 or upgrades to 1.7.8.0, inject into database the Product Page Feature flag

This is what the script `INSERT INTO` does. We need to decide on a way to do it.

#### Make sure all wordings are translatable, including feature flag names and descriptions

#### Modal improvements

- Allow confirmation modal to adapt the wording depending if it's plural or singular
- Only show the modal when flags are enabled, not disabled

#### Link the Product Page V2 flag to V2 Page Product

When we enable/disable Product Page V2 flag this should show/hide the buttons to access Product Page V2




---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/23703)
